File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/Hyun-suck Lee/［PCCP 기출문제］ 1번 ／ 붕대 감기.java
// 이번에는 객체로 풀어보기

class Solution {
    // bandage [ [최대_지속_시간, 초당_회복량, 추가_회복량], ... ]
    // attacks [ [time, damage], ... ]
    public int solution(int[] bandage, int health, int[][] attacks) {
        Hero hero = new Hero(bandage, health);
    
        int maxRound  = attacks[attacks.length - 1][0],
            attackIdx = 0;
        for (int round = 1; round <= maxRound; round++) {
            // 공격 라운드 아님
            if (round != attacks[attackIdx][0]) {
                hero.heal();
                continue;
            }
            // 공격 라운드 맞음
            hero.beAttacked(attacks[attackIdx][1]);
            attackIdx++;
            
            if (hero.isDie()) break;
        }
        
        return hero.getHp();
    }
}

class Hero {
    private int maxHp;
    private int hp;
    private int healTime;       // 현재 힐 지속시간
    private int healAmount;         
    private int rewardTime;     // 추가 회복하는 회복 지속시간
    private int rewardAmount;
    
    public Hero(int[] bandage, int health) {
        this.hp             = health;
        this.maxHp          = health;
        this.healTime       = 0;
        this.rewardTime     = bandage[0];
        this.healAmount     = bandage[1];
        this.rewardAmount   = bandage[2];
    }
    
    public void heal() {
        int healHp   = this.hp + healAmount,
            healTime = this.healTime + 1;
        
        if (healTime == this.rewardTime) {
            healHp += rewardAmount;
            healTime = 0;
        }
        
        this.healTime = healTime;
        this.hp = healHp <= maxHp ? healHp : maxHp;
    }
    
    public void beAttacked(int damage) {
        this.hp -= damage;
        this.healTime = 0;
        
        if (this.hp <= 0) {
            this.hp = -1;
        }
    }
    
    public boolean isDie() {
        return this.hp == -1;
    }
    
    public int getHp() {
        return this.hp;
    }
}

/*
- t(초) 동안 붕대 감으면서 1초마다 x(HP) 회복
- t(초) 연속 회복시 y(HP) 추가 회복
 
- 도중에 공격 당하면 취소
- 공격 당하는 순간에는 회복 불가

- 기술이 끝나거나 공격 당하면 바로 다시 사용

return HP > 0 ? HP : -1;

*/

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/Tae Seong Hong/［PCCP 기출문제］ 1번 ／ 붕대 감기.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(vector<int> bandage, int health, vector<vector<int>> attacks) {
    int answer = 0;
    
    int last_time = attacks[attacks.size() - 1][0];
    int time = 1;
    int max_health = health;
    int heal_time = 0;
    int mob_count = 0;
    
    while(time <= last_time) {
        int attack_time = attacks[mob_count][0];
        int damage = attacks[mob_count][1];
        
        if(time == attack_time) {
            health -= damage;
            heal_time = 0;
            mob_count++;
            if(health <= 0) {
                answer = -1;
                break;
            }
            answer = health;
        } else {
            heal_time++;
            if(heal_time == bandage[0]) {
                health += bandage[2];
                heal_time = 0;
            }
            health += bandage[1];
            if(health >= max_health)
                health = max_health;
        }
        time++;
    }
    
    return answer;
}

/*
t초 동안 1초/x만큼 -> t = 4, x = 4 -> 16
t초 연속 성공 -> y만큼 추가 회복
최대 체력 이상은 회복 불가능
공격 당하면 취소
공격은 정해진 피해량 만큼 체력이 줄고, 0이하면 사망
죽으면 -1을 리턴

bandage = [시전 시간 t, 초당 회복량 x, 추가 회복량 y]
health = 최대 체력
attacks = [[공격 시간 t, 피해량 x]]

끝나는 시간은 몬스터의 마지막 공격

* 풀이
 몬스터의 마지막 공격이 끝나는 시간이 종료 시간이기에 마지막 공격 시간을 뽑아서 반복문 시작
 몹의 처음 공격 시간과 데미지를 저장
 현재 시간이 공격 시간이라면 데미지를 누적, 동시에 붕대 연속 타임(heal_time)을 0으로 초기화
 다음 몹의 공격 시간과 데미지를 저장할 수 있도록 mob_count 증가
 맞아서 피가 0 이하면 answer = -1 저장 후 즉시 종료
 
 그게 아니라면(공격 시간이 아니면),
 일단 붕대를 감는거니까 연속 타임(heal_time)을 1증가
 연속 타임을 만족했는지 확인하고 확인했으면 추가 체력 증가
 붕대 체력만큼 체력 회복 후 체력이 넘치는지 확인하고 넘치면 최대치로 고정
 
 시간 증가 -> 반복
 answer는 몬스터의 공격을 맞고 마지막 체력을 answer에 저장
*/